src/mono/mono/utils/mono-context.c

```diff
@@ -534,10 +534,27 @@ mono_sigctx_to_monoctx (void *sigctx, MonoContext *mctx)
 #endif
 #ifdef __linux__
 	struct fpsimd_context *fpctx = (struct fpsimd_context*)&((ucontext_t*)sigctx)->uc_mcontext.__reserved;
-	int i;
 
-	g_assert (fpctx->head.magic == FPSIMD_MAGIC);
-	for (i = 0; i < 32; ++i)
+	size_t size = 0;
+	do {
+		struct fpsimd_context *fpctx_temp = (struct fpsimd_context*)&(((ucontext_t*)sigctx)->uc_mcontext.__reserved[size]);
+
+		if (fpctx_temp->head.magic == FPSIMD_MAGIC)
+		{
+			g_assert (fpctx_temp->head.size >= sizeof (struct fpsimd_context));
+			g_assert (size + fpctx_temp->head.size <= sizeof (((ucontext_t*)sigctx)->uc_mcontext.__reserved));
+
+			fpctx = fpctx_temp;
+			break;
+		}
+
+		if (fpctx_temp->head.size == 0)
+			break;
+
+		size += fpctx_temp->head.size;
+	} while (size + sizeof (struct fpsimd_context) <= sizeof (((ucontext_t*)sigctx)->uc_mcontext.__reserved));
+
+	for (int i = 0; i < 32; ++i)
```

Comment:
@fanyang-mono if we don't find the FPSIMD_MAGIC, we just write into `mctx->fregs[i] = fpctx->vregs [i]` - while `fpctx` is just pointing to the beginning of `__reserved`?  So we fill it with junk.  That seems a little fragile.  Can we write 0's into `fregs` if we don't find `FPSIMD_MAGIC`?

I think you said in the "arm64 app on x64 Android emultator" scenario we don't see any FPSIMD_MAGIC at all, right? so we should definitely handle that case in a reasonable way instead of just copying some random values into the MonoContext.

Reply:
Oh, yeah. That makes sense.